src/premiumvideodevice/PremiumVideoStreamHandler.ts

```diff
@@ -0,0 +1,211 @@
+   * loading new data
+   */
+  async setInputMediaStream(inputMediaStream: MediaStream | null): Promise<void> {
+    if (!inputMediaStream) {
+      this.stop();
+      return;
+    }
+
+    if (inputMediaStream.getVideoTracks().length === 0) {
+      this.logger.error('No video tracks in input media stream, ignoring');
+      return;
+    }
+
+    this.inputVideoStream = inputMediaStream;
+    const settings = this.inputVideoStream.getVideoTracks()[0].getSettings();
+    this.canvasOutput.width = settings.width;
+    this.canvasOutput.height = settings.height;
+    this.videoInput.addEventListener('loadedmetadata', this.process); // this will be the trigger function
+    this.videoInput.srcObject = this.inputVideoStream;
+    // avoid iOS safari full screen video -- not sure what this does
+    this.videoInput.setAttribute('playsinline', 'true');
+
+    this.videoInput.load();
+    try {
+      await this.videoInput.play();
+    } catch {
+      this.logger.warn('Video element play() overriden by another load()');
+    }
+
+    this.cloneInputAudioTracksToOutput();
+  }
+
+  /**
+   * Transform a singular frame from the input stream to apply desired
+   * video effects. Then place the output image data onto our output canvas
+   * configured with an output stream
+   */
+  process = async (_event: Event): Promise<void> => {
+    if (!this.inputVideoStream) {
+      return;
+    }
+    const processVideoStart = performance.now();
+
+    // Draw the videoInput onto our input canvas
+    if (this.videoInput.videoWidth) {
+      // Match input canvas to same dimensions as our input stream
+      if (
+        this.canvasInput.width !== this.videoInput.videoWidth ||
+        this.canvasInput.height !== this.videoInput.videoHeight
+      ) {
+        this.canvasInput.width = this.videoInput.videoWidth;
+        this.canvasInput.height = this.videoInput.videoHeight;
+      }
+      this.inputCtx.drawImage(this.videoInput, 0, 0);
+    }
+
+    // Collect out input frame data
+    let inputImageData = this.inputCtx.getImageData(
+      0,
+      0,
+      this.canvasInput.width,
+      this.canvasInput.height
+    );
+
+    // Transform input data
+    let transformedImageData = await this.videoEffectDriver.apply(inputImageData);
+
+    // Confirm that the output canvas is still matching video frame size
+    if (
+      this.canvasOutput.width !== this.videoInput.videoWidth &&
+      this.canvasOutput.height !== this.videoInput.videoHeight
+    ) {
+      this.canvasOutput.width = this.videoInput.videoWidth;
+      this.canvasOutput.height = this.videoInput.videoHeight;
+    }
+
+    // Place transformed data onto output canvas
+    this.outputCtx.putImageData(transformedImageData, 0, 0);
+
+    // timing to maintain requested fps
+    const processVideoLatency = performance.now() - processVideoStart;
+    //const leave = (1000 * 2) / this.framerate - processVideoLatency; // half fps
```

Comment:
what is this line for? can we remove it?

Reply:
leave can be used to recognize when processing a singular frame is taking too much time to maintain a certain frame rate. I will take it out for now while we are not concerned with performance throttlers right now